Wake Lock permission denied after visibilitychange

cdumez opened the pull request:
#### d901ad78bdb6536b4c0bdcbf85b1c1e40b8bdfc4
<pre>
Wake Lock permission denied after visibilitychange
<a href="https://bugs.webkit.org/show_bug.cgi?id=255363">https://bugs.webkit.org/show_bug.cgi?id=255363</a>
rdar://108279602

Reviewed by Geoffrey Garen.

Unlike other browser engines, WebKit requires a transient activation to acquire
a screen wake lock. While this is fine for the initial request, this seems a
bit restrictive for follow-up requests, especially considering that the screen
wake lock gets released when the document is no longer visible. It is also
common on the web to try and re-acquire the screen wake lock when the document
becomes visible again (in the visibilitychange event), in which case the JS
doesn&apos;t have transient activation.

To address this issue, we make the authorization from transient activation
sticky. If the page has previously acquired a screen wake lock with transient
activation, we will remember this permission for the lifetime of the document.

* LayoutTests/http/wpt/screen-wake-lock/sticky-permission-after-transient-activation-expected.txt: Added.
* LayoutTests/http/wpt/screen-wake-lock/sticky-permission-after-transient-activation.html: Added.
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::request):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.h:

Canonical link: <a href="https://commits.webkit.org/263382@main">https://commits.webkit.org/263382@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b073b6c0dce9e99b7934e564b2209cbd2d554c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4471 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5702 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4477 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4460 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4318 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4691 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4467 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5696 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1962 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3796 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/6218 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3790 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5395 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4272 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3468 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3794 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1091 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/7872 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4126 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on LayoutTests/http/wpt/screen-wake-lock/sticky-permission-after-transient-activation.html:
```diff
@@ -0,0 +1,21 @@
+    await lock.release();
+
+    // This document was previously authorized so it should be able to require the wake lock again.
+    lock = await navigator.wakeLock.request('screen');
```
Maybe overkill, but: 

```suggestion
    assert_true(navigator.userActivation.hasBeenActive, "has sticky activation");
    lock = await navigator.wakeLock.request('screen');
```

marcoscaceres commented on Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
```diff
@@ -67,9 +67,15 @@ void WakeLock::request(WakeLockType lockType, Ref<DeferredPromise>&& promise)
-            permission = hasTransientActivation ? PermissionState::Granted : PermissionState::Denied;
+    PermissionController::shared().query(document->clientOrigin(), PermissionDescriptor { PermissionName::ScreenWakeLock }, *document->page(), PermissionQuerySource::Window, [this, protectedThis = Ref { *this }, document = Ref { *document }, hasTransientActivation, promise = WTFMove(promise), lockType](std::optional<PermissionState> permission) mutable {
+        if (!permission || *permission == PermissionState::Prompt) {
+            if (hasTransientActivation || m_wasPreviouslyAuthorizedDueToTransientActivation) {
```
Do we need m_wasPreviouslyAuthorizedDueToTransientActivation? can't we check if window->hasStickyActivation?

cdumez replied:
Wouldn't your proposal be a LOT more permissive?
Basically, if I have ever interacted with the document, then the JS will be able to acquire screen wake lock whenever?

I was trying to maintain our current behavior of requiring a transient activation for the initial request.

marcoscaceres replied:
That's true, at the same time. I'm trying to see if we can avoid inventing a "new(ish)" user activation model... but we might need one here 🤔

geoffreygaren replied:
I wonder if other permissions that depend on sticky activation could reasonably migrate to this model. In theory, in a world of many permissions, it feels bad that a single click or tap might flood the user with many enabled-by-activation behaviors.

The right answer will probably depend on an empirical analysis of how certain permissions are used in practice.

marcoscaceres replied:
I think there are very few things that depend on sticky, so it might be worth doing. I'll look into that from the HTML side.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Wake Lock permission denied after visibilitychange
https://bugs.webkit.org/show_bug.cgi?id=255363
rdar://108279602

Reviewed by Geoffrey Garen.

Unlike other browser engines, WebKit requires a transient activation to acquire
a screen wake lock. While this is fine for the initial request, this seems a
bit restrictive for follow-up requests, especially considering that the screen
wake lock gets released when the document is no longer visible. It is also
common on the web to try and re-acquire the screen wake lock when the document
becomes visible again (in the visibilitychange event), in which case the JS
doesn't have transient activation.

To address this issue, we make the authorization from transient activation
sticky. If the page has previously acquired a screen wake lock with transient
activation, we will remember this permission for the lifetime of the document.

* LayoutTests/http/wpt/screen-wake-lock/sticky-permission-after-transient-activation-expected.txt: Added.
* LayoutTests/http/wpt/screen-wake-lock/sticky-permission-after-transient-activation.html: Added.
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::request):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.h:

Canonical link: https://commits.webkit.org/263382@main
